src/hypothesis/strategies.py

```diff
@@ -713,65 +714,77 @@ def characters(whitelist_categories=None, blacklist_categories=None,
+      exception.
+
+    The ``_codepoint`` arguments must be integers between zero and
+    :obj:`python:sys.max_unicode`.  The ``_characters`` arguments must be
+    collections of length-one unicode strings, such as a unicode string.
+
+    The ``_categories`` arguments must be collections of strings (type str
+    on any version) identifying a Unicode `general category`_.  For example,
+    ``('Nd', 'Lu')`` signifies "Number, decimal digit" and "Letter, uppercase".
+    A single letter ('major category') can be given to match all corresponding
+    categories, for example ``'P'`` for characters in any punctuation category.
+
+    .. _general category: https://wikipedia.org/wiki/Unicode_character_property
+
+    Examples from this strategy shrink towards the codepoint for ``'0'``,
+    or the first allowable codepoint after it if ``'0'`` is excluded.
 
     """
-    if (
-        min_codepoint is not None and max_codepoint is not None and
-        min_codepoint > max_codepoint
-    ):
-        raise InvalidArgument(
-            'Cannot have min_codepoint=%d > max_codepoint=%d ' % (
-                min_codepoint, max_codepoint
-            )
-        )
+    check_valid_size(min_codepoint, 'min_codepoint')
+    check_valid_size(max_codepoint, 'max_codepoint')
+    check_valid_interval(min_codepoint, max_codepoint,
+                         'min_codepoint', 'max_codepoint')
     if all((whitelist_characters is not None,
             min_codepoint is None,
             max_codepoint is None,
             whitelist_categories is None,
             blacklist_categories is None,
             )):
         raise InvalidArgument(
-            'Cannot have just whitelist_characters=%r alone, '
-            'it would have no effect. Perhaps you want sampled_from()' % (
-                whitelist_characters,
-            )
-        )
-    if (
-        whitelist_characters is not None and
-        blacklist_characters is not None and
-        set(blacklist_characters).intersection(set(whitelist_characters))
-    ):
+            'Passing only whitelist_characters=%r would have no effect. '
+            'Perhaps you want sampled_from() ?' % (whitelist_characters,))
+    blacklist_characters = blacklist_characters or ''
+    whitelist_characters = whitelist_characters or ''
+    overlap = set(blacklist_characters).intersection(whitelist_characters)
+    if overlap:
         raise InvalidArgument(
             'Characters %r are present in both whitelist_characters=%r, and '
             'blacklist_characters=%r' % (
-                set(blacklist_characters).intersection(
-                    set(whitelist_characters)
-                ),
-                whitelist_characters, blacklist_characters,
-            )
-        )
+                sorted(overlap), whitelist_characters, blacklist_characters))
+    blacklist_categories = as_general_categories(
+        blacklist_categories, 'blacklist_categories')
+    whitelist_categories = as_general_categories(
```

Comment:
Have we thought about the fact that this could change the behavior for some users who were previously (incorrectly) using one-letter major categories (and don't get exceptions either before or now)?

Yours:

```
In [64]: characters(whitelist_categories=['L'], whitelist_characters=['a','b']).example()
Out[64]: 'b'
```

Master:

```
In [2]: characters(whitelist_categories=['L'], whitelist_characters=['a','b']).example()
Out[2]: 'ď'

In [3]: characters(whitelist_categories=['L'], whitelist_characters=['a','b']).example()
Out[3]: '鹼'
```

Seems like an extremely rare (maybe non-existent in real code) edge case, so I'm definitely not saying it's a show-stopper, but some things I think I would want before this happens:

- make sure the decision is made consciously
- someone with more than 1 week of experience with how Hypothesis thinks about backwards compatibility (i.e., not me) should "review"/approve that decision (I don't need anyone else to review the code in the PR -- just that decision)
- more explicitly call it out in the changelog, something like `One-letter characters (which were previously silently ignored) are interpreted as Unicode major classes if possible (and still silently ignored if not).`

Reply:
Oh, and of course besides being a very rare-seeming case, probably the user would be *better off* with the new behavior. The old behavior is a silent error that's not doing what they intended. So I think this is probably fine -- I'm just not comfortable being the one to say so.

Reply:
- Yep, deliberate decision.
- Under our deprecation rules, anything not documented is not public and thus subject to change - our prior docs about this were extremely sparse, but I think it's reasonable to clarify them and support this new case at the same time.  (Ping @DRMacIver to agree or not on this point?)
- I think the changelog is sufficient, because it clearly describes the supported (and non-deprecated) behavior.

Reply:
Cool, I'll withdraw my request for a changelog change. (Still just too new here to be happy giving the decision 👍 myself, but that's not coming from disagreement at all. Just new.)

Reply:
I agree the change is fine, but I also agree with @dchudz that it's worth calling out explicitly in the changelog that this is a departure from what would previously have happened - no harm in doing so, and if it results in someone running into new issues it helps clarify the "What changed?" head scratcher.